tools/mpy_ld.py

```diff
@@ -590,40 +670,75 @@ def do_relocation_text(env, text_addr, r):
             return
         assert 0
 
+    elif env.arch.name == "EM_RISCV" and r_info_type in (
+        R_RISCV_TLS_GD_HI20,
+        R_RISCV_TLSDESC_HI20,
+        R_RISCV_TLSDESC_ADD_LO12,
+        R_RISCV_TLSDESC_CALL,
+    ):
+        # TLS relocations are not supported.
+        raise LinkError("{}: RISC-V TLS relocation: {}".format(s.filename, s.name))
+
+    elif env.arch.name == "EM_RISCV" and r_info_type in (
+        R_RISCV_TPREL_HI20,
+        R_RISCV_TPREL_LO12_I,
+        R_RISCV_TPREL_LO12_S,
+        R_RISCV_TPREL_ADD,
+    ):
+        # ThreadPointer-relative relocations are not supported.
+        raise LinkError("{}: RISC-V TP-relative relocation: {}".format(s.filename, s.name))
+
+    elif env.arch.name == "EM_RISCV" and r_info_type in (R_RISCV_SET_ULEB128, R_RISCV_SUB_ULEB128):
+        # 128-bit value relocations are not supported
+        raise LinkError("{}: RISC-V ULEB128 relocation: {}".format(s.filename, s.name))
+
+    elif env.arch.name == "EM_RISCV" and r_info_type in (R_RISCV_RELAX, R_RISCV_ALIGN):
+        # To keep things simple, no relocations are relaxed and thus no
+        # size optimisation is performed even if there is the chance, along
+        # with no offsets to fix up.
+        return
+
+    elif env.arch.name == "EM_RISCV":
+        (addr, value) = process_riscv32_relocation(env, text_addr, r)
+
     else:
         # Unknown/unsupported relocation
         assert 0, r_info_type
 
-    # Write relocation
-    if reloc_type == "le32":
-        (existing,) = struct.unpack_from("<I", env.full_text, r_offset)
-        struct.pack_into("<I", env.full_text, r_offset, (existing + reloc) & 0xFFFFFFFF)
-    elif reloc_type == "thumb_b":
-        b_h, b_l = struct.unpack_from("<HH", env.full_text, r_offset)
-        existing = (b_h & 0x7FF) << 12 | (b_l & 0x7FF) << 1
-        if existing >= 0x400000:  # 2's complement
-            existing -= 0x800000
-        new = existing + reloc
-        b_h = (b_h & 0xF800) | (new >> 12) & 0x7FF
-        b_l = (b_l & 0xF800) | (new >> 1) & 0x7FF
-        struct.pack_into("<HH", env.full_text, r_offset, b_h, b_l)
-    elif reloc_type == "xtensa_l32r":
-        l32r = unpack_u24le(env.full_text, r_offset)
-        assert l32r & 0xF == 1  # RI16 encoded l32r
-        l32r_imm16 = l32r >> 8
-        l32r_imm16 = (l32r_imm16 + reloc >> 2) & 0xFFFF
-        l32r = l32r & 0xFF | l32r_imm16 << 8
-        pack_u24le(env.full_text, r_offset, l32r)
-    else:
-        assert 0, reloc_type
+    if env.arch.name != "EM_RISCV":
```

Comment:
To minimise diff and make it easier to change in the future, I suggest:
```py
if env.arch.name == "EM_RISCV":
    # (comment why we don't do anything here)
    pass
elif reloc_type == "le32":
    ...
```

Reply:
Makes sense, I've updated the PR to reflect this suggestion.  Thanks!